Account/FxADeviceRegistrator.swift

```diff
@@ -22,39 +22,74 @@ public enum FxADeviceRegistratorError: MaybeErrorType {
     case AccountDeleted
     case CurrentDeviceNotFound
     case InvalidSession
+    case UnknownDevice
 
     public var description: String {
         switch self {
         case AccountDeleted: return "Account no longer exists."
         case CurrentDeviceNotFound: return "Current device not found."
         case InvalidSession: return "Session token was invalid."
+        case UnknownDevice: return "Unknown device."
         }
     }
 }
 
+public class FxADeviceRegistration: NSObject, NSCoding {
+    /// The device identifier identifying this device.  A device is uniquely identified
+    /// across the lifetime of a Firefox Account.
+    let id: String
+
+    /// The version of the device registration. We use this to re-register
+    /// devices after we update what we send on device registration.
+    let version: Int
+
+    /// The last time we successfully (re-)registered with the server.
+    let lastRegistered: Timestamp
+
+    init(id: String, version: Int, lastRegistered: Timestamp) {
+        self.id = id
+        self.version = version
+        self.lastRegistered = lastRegistered
+    }
+
+    public convenience required init(coder: NSCoder) {
+        let id = coder.decodeObjectForKey("id") as! String
+        let version = coder.decodeObjectForKey("version") as! Int
+        let lastRegistered = (coder.decodeObjectForKey("lastRegistered") as! NSNumber).unsignedLongLongValue
+        self.init(id: id, version: version, lastRegistered: lastRegistered)
+    }
+
+    public func encodeWithCoder(aCoder: NSCoder) {
+        aCoder.encodeObject(id, forKey: "id")
+        aCoder.encodeObject(version, forKey: "version")
+        aCoder.encodeObject(NSNumber(unsignedLongLong: lastRegistered), forKey: "lastRegistered")
+    }
+}
+
 public class FxADeviceRegistrator {
-    public static func registerOrUpdateDevice(account: FirefoxAccount, state: MarriedState, client: FxAClient10? = nil) -> Deferred<Maybe<FxADeviceRegistrationResult>> {
-        if let _ = account.fxaDeviceId where account.deviceRegistrationVersion == DeviceRegistrationVersion {
-            return deferMaybe(FxADeviceRegistrationResult.AlreadyRegistered)
+    public static func registerOrUpdateDevice(account: FirefoxAccount, sessionToken: NSData, client: FxAClient10? = nil) -> Deferred<Maybe<FxADeviceRegistrationResult>> {
+        if let registration = account.deviceRegistration where registration.version == DeviceRegistrationVersion &&
```

Comment:
I find this line hard to read.  I feel the comment is misplaced, as well -- it doesn't describe the branch it is in.